torch_xla/experimental/xla_quantized_matmul.py

```diff
@@ -21,83 +21,132 @@ def _check_per_channel_quant_weight_dtype_shapes(input_dim, output_dim, w,
+    # Per-channel quant.
+    _check_per_channel_quant_weight_dtype_shapes(x.shape[-1], scaler.shape[0],
+                                                 w, scaler)
+    return F.linear(x, w) * scaler
+  else:
+    # Blockwise quant.
+    _check_blockwise_quant_weight_dtype_shapes(x.shape[-1], w.shape[-1],
+                                               block_size, w, scaler)
+    x = x.reshape(*x.shape[:-1], x.shape[-1] // block_size, block_size)
+    out = torch.einsum('scn,...sc->...sn', w, x)
+    out = torch.einsum('sn,...sn->...n', scaler, out)
+    return out
 
 
 @impl(XLA_LIB, "quantized_matmul", "CompositeExplicitAutograd")
 def quantized_matmul(x: torch.Tensor,
                      w: torch.Tensor,
                      scaler: torch.Tensor,
-                     blocksize: int = -1,
+                     block_size: int = -1,
                      int4_weight: bool = False):
-  assert blocksize == -1, "blockwise quantization is not supported yet."
-  # Per-channel quant.
-  _check_per_channel_quant_weight_dtype_shapes(x.shape[-1], scaler.shape[0], w,
-                                               scaler)
-  w = w.to(x.dtype)
-  return torch.mul(F.linear(x, w), scaler)
+  if block_size == -1:
+    # Per-channel quant.
+    _check_per_channel_quant_weight_dtype_shapes(x.shape[-1], scaler.shape[0],
+                                                 w, scaler)
+    w = w.to(x.dtype)
+    return torch.mul(F.linear(x, w), scaler)
+  else:
+    # Blockwise quant.
+    _check_blockwise_quant_weight_dtype_shapes(x.shape[-1], w.shape[-1],
+                                               block_size, w, scaler)
+    x = x.reshape(*x.shape[:-1], x.shape[-1] // block_size, block_size)
+    w = w.to(x.dtype)
+    out = torch.einsum('scn,...sc->...sn', w, x)
+    out = torch.einsum('sn,...sn->...n', scaler, out)
+    return out
 
 
 class XlaQuantizedLinear(torch.nn.Module):
 
   def __init__(self,
                input_dim,
                output_dim,
-               blocksize=-1,
+               block_size=-1,
                int4_weight: bool = False):
     super().__init__()
-    assert blocksize == -1, "Only per-channel quantization is supported."
     self.input_dim = input_dim
     self.output_dim = output_dim
-    self.blocksize = blocksize
+    self.block_size = block_size
     self.int4_weight = int4_weight
     self.register_buffer('weight',
                          torch.zeros(output_dim, input_dim).to(torch.int8))
     self.register_buffer('weight_scaler', torch.zeros(output_dim))
 
   def load_quantized_weight(self, weight, weight_scaler):
     '''
-    Weight shape: [output_channel, input_channel]
-    Weight scaler shape: [output_channel]
+    Weight shape:
+      per-channel quant: [out_channel, in_channel].
+      block_wise quant: [in_channel / block_size, block_size, out_channel].
+
+    Weight scaler shape:
+      per-channel quant: [out_channel,].
+      blockwise quant: [in_channel / block_size, out_channel].
```

Comment:
nit, can fix in a follow up. I find it a bit confusing to use `wegiht shape` here because I thought it is a paramter. I think it is better to follow pytorch documentation pattern
```
weight (Tensor): for per-channel quant, it should be a tensor with shape [out_channel, in_channel] .....
```

Reply:
sounds good, let me update.